quantus/evaluation.py

```diff
@@ -87,43 +99,58 @@ def evaluate(
             "to each input."
         )
 
+        if explain_func_kwargs is None:
+            explain_func_kwargs = {}
+
         for method in xai_methods:
 
             results[method] = {}
-
-            for metric, metric_func in metrics.items():
-
-                if progress:
-                    print(f"Evaluating {method} explanations on {metric} metric...")
-
-                results[method][metric] = agg_func(
-                    metric_func(
-                        model=model,
-                        x_batch=x_batch,
-                        y_batch=y_batch,
-                        a_batch=a_batch,
-                        s_batch=s_batch,
-                        explain_func_kwargs={
-                            **explain_func_kwargs,
-                            **{"method": method},
-                        },
-                        **call_kwargs,
+            explain_funcs[method] = explain
+
+            for (metric, metric_func) in metrics.items():
+
+                results[method][metric] = {}
+
+                for (call_kwarg_str, call_kwarg) in call_kwargs.items():
+
+                    if progress:
+                        print(
+                            f"Evaluating {method} explanations on {metric} metric on set of call parameters {call_kwarg_str}..."
+                        )
+
+                    results[method][metric][call_kwarg_str] = agg_func(
+                        metric_func(
+                            model=model,
+                            x_batch=x_batch,
+                            y_batch=y_batch,
+                            a_batch=a_batch,
+                            s_batch=s_batch,
+                            explain_func=explain_funcs[method],
+                            explain_func_kwargs={
+                                **explain_func_kwargs,
+                                **{"method": method},
```

Comment:
I find it quite confusing here that each element of "xai_methods" is basically used both as a key for the results as well as to determine the XAI method...

this option to pass "xai_methods" as a list is a bit redundant (since Dict[str, Dict] does the same but with more control over parameters), makes kwarg variation quite tedious (as a user would need an extra outside loop to iterate "explain_func_kwargs") and thus encourages application of default value explanations (leading to bad comparisons in literature, such as e.g. "Occlusion vs. IntegratedGradients" instead of "occlusion_with_window_size_2 vs. IG_with_10_steps"). It also makes the evaluate function confusing, since explain_func_kwargs won't do anything for the Dict[str, Dict] option. It also simply won't work if the user uses zennit (and the quantus explanation_func wrapper around it) for explanations.

I think the List[str] option should be removed from xai_methods, because it behaves a bit erratically and differently to the other two options and does not really add much usability.

Reply:
Thanks @leanderweber. I see your point. We can leave the single option to pass Dict[str, Dict] as xai_methods, since all the other options have similar problems. For Dict[str, np.ndarray] option in xai_methods, there is no way to indicate which arguments were used for generation of the array, but they might be needed for metric calculation (right now it just uses the same "explain_func_kwargs" for every entry in the dict). The option Dict[str, Callable] is basically the same as List(str).